### ISSUE TITLE
WASI apps are too big

### ISSUE THREAD

**Issue by mhmd-azeez** [dotnet/runtime#94286]
Repro:

```
 dotnet new wasiconsole -o SizeApp
```

```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>net8.0</TargetFramework>
    <RuntimeIdentifier>wasi-wasm</RuntimeIdentifier>
    <OutputType>Exe</OutputType>
    <PublishTrimmed>true</PublishTrimmed>

    <WasmSingleFileBundle>true</WasmSingleFileBundle>
    <InvariantGlobalization>true</InvariantGlobalization>
    <TrimMode>full</TrimMode>
    <DebuggerSupport>false</DebuggerSupport>
    <EventSourceSupport>false</EventSourceSupport>
    <StackTraceSupport>false</StackTraceSupport>
    <UseSystemResourceKeys>true</UseSystemResourceKeys>
    <NativeDebugSymbols>false</NativeDebugSymbols>
  </PropertyGroup>
</Project>
```

When using `dotnet build` for both Release & Debug configurations, a ~23mb .wasm file is generated called `SizeApp.wasm`.
When using `dotnet publish` a 10mb .wasm file is generated called `dotnet.wasm` (but this file can't be run using wasmtime directly):
```
PS D:\x\dotnet\SizeApp> wasmtime run .\bin\Release\net8.0\wasi-wasm\publish\dotnet.wasm
'/managed' in MONO_PATH doesn't exist or has wrong permissions.
Failed to load ./icudt.dat
Error: failed to run main module `.\bin\Release\net8.0\wasi-wasm\publish\dotnet.wasm`

Caused by:
    0: failed to invoke command default
    1: error while executing at wasm backtrace:
           0: 0x2753d3 - <unknown>!fseek
           1: 0x5ad4 - <unknown>!load_icu_data
           2: 0x5bca - <unknown>!mono_wasm_load_runtime
           3: 0x6852 - <unknown>!main
           4: 0x272e08 - <unknown>!__main_void
           5: 0x5911 - <unknown>!_start
           6: 0x285fd9 - <unknown>!_start.command_export
       note: using the `WASMTIME_BACKTRACE_DETAILS=1` environment variable may show more debugging information
    2: wasm trap: uninitialized element
```

[This [comment](https://github.com/dotnet/runtime/issues/65895#issuecomment-1780600554) ](https://github.com/dotnet/runtime/issues/65895#issuecomment-1780600554) suggests that using the above configuration the size should come down to ~3mb.

More details:
```
PS D:\x\dotnet\SizeApp> dotnet --info
.NET SDK:
 Version:   8.0.100-rc.2.23502.2
 Commit:    0abacfc2b6

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.22621
 OS Platform: Windows
 RID:         win-x64
 Base Path:   C:\Program Files\dotnet\sdk\8.0.100-rc.2.23502.2\

.NET workloads installed:
 [wasi-experimental]
   Installation Source: SDK 8.0.100-rc.2
   Manifest Version:    8.0.0-rc.2.23479.6/8.0.100-rc.2
   Manifest Path:       C:\Program Files\dotnet\sdk-manifests\8.0.100-rc.2\microsoft.net.workload.mono.toolchain.current\8.0.0-rc.2.23479.6\WorkloadManifest.json
   Install Type:              Msi


Host:
  Version:      8.0.0-rc.2.23479.6
  Architecture: x64
  Commit:       0b25e38ad3

.NET SDKs installed:
  7.0.403 [C:\Program Files\dotnet\sdk]
  8.0.100-rc.2.23502.2 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 6.0.24 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.13 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 8.0.0-rc.2.23480.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 6.0.24 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.13 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 8.0.0-rc.2.23479.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 6.0.24 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 7.0.13 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 8.0.0-rc.2.23479.10 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

Other architectures found:
  x86   [C:\Program Files (x86)\dotnet]
    registered at [HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\x86\InstallLocation]

Environment variables:
  Not set

global.json file:
  Not found

Learn more:
  https://aka.ms/dotnet/info

Download .NET:
  https://aka.ms/dotnet/download
```

```
PS D:\x\dotnet\SizeApp> dotnet workload list  

Installed Workload Id      Manifest Version                     Installation Source
-----------------------------------------------------------------------------------
wasi-experimental          8.0.0-rc.2.23479.6/8.0.100-rc.2      SDK 8.0.100-rc.2

Use `dotnet workload search` to find additional workloads to install.
```

Complete repro:
[SizeApp.zip](https://github.com/dotnet/runtime/files/13236438/SizeApp.zip)

**Comment by maraf** [comment]
Thanks for the details!

The build/publish output is in the `AppBundle` folder (eg. `.\bin\Release\net8.0\wasi-wasm\AppBundle`). 
In case of `WasmSingleFileBundle=true` the wasm file is named after project (eg. `SizeApp.wasm`) (I mentioned it incorrectly in the original issue, sorry for that).

**Comment by mhmd-azeez** [comment]
No problem, looking forward to a workaround for the size issue. I am waiting on that before publishing a blog post about WASI & .NET

**Comment by mhmd-azeez** [comment]
@maraf that's strange, this is what I get:
```
PS D:\x\dotnet\SizeApp> dotnet build -c Release
MSBuild version 17.8.0+6cdef4241 for .NET
  Determining projects to restore...
  All projects are up-to-date for restore.
C:\Program Files\dotnet\sdk\8.0.100-rc.2.23502.2\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInference.targets(311,5): message NETSDK1057: You are using a preview vers
ion of .NET. See: https://aka.ms/dotnet-support-policy [D:\x\dotnet\SizeApp\SizeApp.csproj]
  SizeApp -> D:\x\dotnet\SizeApp\bin\Release\net8.0\wasi-wasm\SizeApp.dll
  MicrosoftNetCoreAppRuntimePackRidNativeDir: C:\Program Files\dotnet\packs\Microsoft.NETCore.App.Runtime.Mono.wasi-wasm\8.0.0-rc.2.23479.6\runtimes\wasi-wasm\native\
  Performing WASI SDK build: "D:\tools\wasi_sdk\bin\clang.exe" "@D:\x\dotnet\SizeApp\obj\Release\net8.0\wasi-wasm\wasm\for-build\clang-compile.rsp"

Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:12.02
PS D:\x\dotnet\SizeApp> ls .\bin\Release\net8.0\wasi-wasm\AppBundle\

    Directory: D:\x\dotnet\SizeApp\bin\Release\net8.0\wasi-wasm\AppBundle

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
d----           11/2/2023  1:17 PM                tmp
-a---           11/2/2023  1:17 PM             30 run-wasmtime.sh
-a---           11/2/2023  1:17 PM           1147 runtimeconfig.bin
-a---           11/2/2023  1:17 PM           1821 SizeApp.runtimeconfig.json
-a---           11/2/2023  1:17 PM       23687753 SizeApp.wasm
```

**Comment by maraf** [comment]
Use publish please. Some of the size savings happend only during publish (I don't know of head which), but a lot of optimizations happen throughout the whole .NET only during publish

**Comment by mhmd-azeez** [comment]
The problem is, publish doesn't produce a `SizeApp.wasm`:
```
PS D:\x\dotnet\SizeApp> dotnet publish -c Release
MSBuild version 17.8.0+6cdef4241 for .NET
  Determining projects to restore...
  All projects are up-to-date for restore.
C:\Program Files\dotnet\sdk\8.0.100-rc.2.23502.2\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInference.targets(311,5): message NETSDK1057: You are using a preview vers
ion of .NET. See: https://aka.ms/dotnet-support-policy [D:\x\dotnet\SizeApp\SizeApp.csproj]
  SizeApp -> D:\x\dotnet\SizeApp\bin\Release\net8.0\wasi-wasm\SizeApp.dll
  Optimizing assemblies for size. This process might take a while.
  SizeApp -> D:\x\dotnet\SizeApp\bin\Release\net8.0\wasi-wasm\publish\
  SizeApp -> D:\x\dotnet\SizeApp\bin\Release\net8.0\wasi-wasm\SizeApp.dll
  SizeApp -> D:\x\dotnet\SizeApp\bin\Release\net8.0\wasi-wasm\publish\
  MicrosoftNetCoreAppRuntimePackRidNativeDir: C:\Program Files\dotnet\packs\Microsoft.NETCore.App.Runtime.Mono.wasi-wasm\8.0.0-rc.2.23479.6\runtimes\wasi-wasm\native\
  Performing WASI SDK build: "D:\tools\wasi_sdk\bin\clang.exe" "@D:\x\dotnet\SizeApp\obj\Release\net8.0\wasi-wasm\wasm\for-publish\clang-compile.rsp"
PS D:\x\dotnet\SizeApp> ls .\bin\Release\net8.0\wasi-wasm\publish\

    Directory: D:\x\dotnet\SizeApp\bin\Release\net8.0\wasi-wasm\publish

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
-a---           9/29/2023  6:56 PM       11190025 dotnet.wasm
-a---           9/29/2023  6:46 PM          20161 driver.c
-a---           9/29/2023  6:46 PM           1381 driver.h
-a---           9/29/2023  6:46 PM           2031 gc-common.h
-a---           9/18/2023  2:24 PM         956416 icudt_CJK.dat
-a---           9/18/2023  2:24 PM         550832 icudt_EFIGS.dat
-a---           9/18/2023  2:24 PM         500480 icudt_hybrid.dat
-a---           9/18/2023  2:24 PM        1107168 icudt_no_CJK.dat
-a---           9/18/2023  2:24 PM        1526128 icudt.dat
-a---           9/18/2023  2:25 PM            454 libicudata.a
-a---           9/18/2023  2:25 PM        3400404 libicui18n.a
-a---           9/18/2023  2:25 PM        2193544 libicuuc.a
-a---           9/29/2023  6:49 PM         962450 libmono-component-debugger-static.a
-a---           9/29/2023  6:49 PM          67758 libmono-component-debugger-stub-static.a
-a---           9/29/2023  6:49 PM          63118 libmono-component-diagnostics_tracing-static.a
-a---           9/29/2023  6:49 PM          23028 libmono-component-diagnostics_tracing-stub-static.a
-a---           9/29/2023  6:49 PM         229738 libmono-component-hot_reload-static.a
-a---           9/29/2023  6:49 PM          40634 libmono-component-hot_reload-stub-static.a
-a---           9/29/2023  6:49 PM         169744 libmono-component-marshal-ilgen-static.a
-a---           9/29/2023  6:49 PM          62306 libmono-component-marshal-ilgen-stub-static.a
-a---           9/29/2023  6:50 PM        1305800 libmono-ee-interp.a
-a---           9/29/2023  6:49 PM          86964 libmono-icall-table.a
-a---           9/29/2023  6:49 PM           2318 libmono-wasm-nosimd.a
-a---           9/29/2023  6:49 PM          27998 libmono-wasm-simd.a
-a---           9/29/2023  6:50 PM       13341326 libmonosgen-2.0.a
-a---           9/29/2023  6:52 PM          49554 libSystem.Globalization.Native.a
-a---           9/29/2023  6:52 PM          88676 libSystem.IO.Compression.Native.a
-a---           9/29/2023  6:52 PM          37652 libSystem.Native.a
-a---           9/29/2023  6:46 PM           1921 main.c
-a---           9/29/2023  6:46 PM           1414 pinvoke.c
-a---           9/29/2023  6:46 PM           1136 pinvoke.h
-a---           11/2/2023  2:14 PM           4967 SizeApp.deps.json
-a---           11/2/2023  2:14 PM           4608 SizeApp.dll
-a---           11/2/2023  1:17 PM           1724 SizeApp.runtimeconfig.json
-a---           9/29/2023  6:46 PM           1000 stubs.c
-a---           9/29/2023  6:46 PM           2225 synthetic-pthread.c
-a---           11/2/2023  2:14 PM          17408 System.Console.dll
-a---           11/2/2023  2:14 PM         840192 System.Private.CoreLib.dll
-a---           9/29/2023  6:56 PM         610748 wasm-bundled-timezones.a
-a---           9/29/2023  6:56 PM            460 wasm-config.h
```

Can you please give me the full list of commands you ran to get the 3mb wasm file?

**Comment by maraf** [comment]
Are you looking into the `AppBundle` folder (even for publish)?

**Comment by mhmd-azeez** [comment]
@maraf Ah! it has indeed produced a ~2.7mb wasm file:
```
PS D:\x\dotnet\SizeApp> ls .\bin\Release\net8.0\wasi-wasm\AppBundle\

    Directory: D:\x\dotnet\SizeApp\bin\Release\net8.0\wasi-wasm\AppBundle

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
d----           11/2/2023  1:17 PM                tmp
-a---           11/2/2023  2:14 PM             30 run-wasmtime.sh
-a---           11/2/2023  1:17 PM           1147 runtimeconfig.bin
-a---           11/2/2023  2:14 PM           1821 SizeApp.runtimeconfig.json
-a---           11/2/2023  2:14 PM        2744828 SizeApp.wasm
PS D:\x\dotnet\SizeApp> wasmtime run .\bin\Release\net8.0\wasi-wasm\AppBundle\SizeApp.wasm
Hello, Wasi Console!
```

Thanks! seems to be working, I was just not looking at the `AppBundle` folder

**Comment by maraf** [comment]
I have opened as new issue for printing the correct output path https://github.com/dotnet/runtime/issues/94291.
I'm going to close this one as resolved. Feel free to reopen it or create a new one if you have more questions
